From 13a6870afd7960e3250a2720f2118afb7116cd4d Mon Sep 17 00:00:00 2001
From: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Date: Wed, 19 May 2021 13:46:36 +0300
Subject: [PATCH] HACK: Make MSI-X work for Xilinx NWL driver

It seems MSI-X doesn't work because the NWL Linux driver doesn't
support it.

MSI doesn't work with the current Linux driver on QEMU because the
driver uses a remapping feature we don't yet support in QEMU.

But with the attached patch to the Linux driver it works for me.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Suggested-by: Edgar E. Iglesias <edgar.iglesias@xilinx.com>
---
 drivers/pci/controller/pcie-xilinx-nwl.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/pci/controller/pcie-xilinx-nwl.c b/drivers/pci/controller/pcie-xilinx-nwl.c
index f3cf7d61924f..9c2cbcd07c1a 100644
--- a/drivers/pci/controller/pcie-xilinx-nwl.c
+++ b/drivers/pci/controller/pcie-xilinx-nwl.c
@@ -443,7 +443,7 @@ static struct irq_chip nwl_msi_irq_chip = {
 
 static struct msi_domain_info nwl_msi_domain_info = {
 	.flags = (MSI_FLAG_USE_DEF_DOM_OPS | MSI_FLAG_USE_DEF_CHIP_OPS |
-		  MSI_FLAG_MULTI_PCI_MSI),
+		  MSI_FLAG_MULTI_PCI_MSI | MSI_FLAG_PCI_MSIX),
 	.chip = &nwl_msi_irq_chip,
 };
 #endif
@@ -451,7 +451,7 @@ static struct msi_domain_info nwl_msi_domain_info = {
 static void nwl_compose_msi_msg(struct irq_data *data, struct msi_msg *msg)
 {
 	struct nwl_pcie *pcie = irq_data_get_irq_chip_data(data);
-	phys_addr_t msi_addr = pcie->phys_pcie_reg_base;
+	phys_addr_t msi_addr = 0xFE440000;
 
 	msg->address_lo = lower_32_bits(msi_addr);
 	msg->address_hi = upper_32_bits(msi_addr);
@@ -616,10 +616,12 @@ static int nwl_pcie_enable_msi(struct nwl_pcie *pcie)
 	nwl_bridge_writel(pcie, nwl_bridge_readl(pcie, I_MSII_CONTROL) |
 			  MSII_STATUS_ENABLE, I_MSII_CONTROL);
 
+#if 0
 	/* setup AFI/FPCI range */
 	base = pcie->phys_pcie_reg_base;
 	nwl_bridge_writel(pcie, lower_32_bits(base), I_MSII_BASE_LO);
 	nwl_bridge_writel(pcie, upper_32_bits(base), I_MSII_BASE_HI);
+#endif
 
 	/*
 	 * For high range MSI interrupts: disable, clear any pending,
-- 
2.31.1

